### PR TITLE
Add support for composite sorting

### PR DESCRIPTION
**Pull request by frankjtao**
Support for the composite sort e.g. http://hapi.fhir.org/baseR4/Observation?_sort=component-code-value-quantity

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 1]
```diff
@@ -542,6 +558,37 @@ private void createSort(QueryStack theQueryStack, SortSpec theSort) {
...
+	private void createCompositeSort(QueryStack theQueryStack, String theResourceName, RestSearchParameterTypeEnum theParamType, String theParamName, boolean theAscending) {
+
+		switch (theParamType) {
+		case STRING:
+			theQueryStack.addSortOnString(myResourceName, theParamName, theAscending);
+			break;
+		case DATE:
+			theQueryStack.addSortOnDate(myResourceName, theParamName, theAscending);
+			break;
+		case REFERENCE:
```
Comment (IanMMarshall) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:571]: I am wondering if it's appropriate to support sorting on composites with `reference`,  `number` and `uri` param types given that the search functionality does not appear to support searching on composites that include these param types. 

If we are going to support these types, then I think we should include tests for these types.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 2]
```diff
@@ -542,6 +558,37 @@ private void createSort(QueryStack theQueryStack, SortSpec theSort) {
...
+			theQueryStack.addSortOnResourceLink(myResourceName, theParamName, theAscending);
+			break;
```
Comment (IanMMarshall) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:573]: One way you could test these edge cases (assuming you keep these) is to create resources with number, reference and/or uri values and then create a custom SearchParameter. The `FhirResourceDaoR4SearchCustomSearchParamTest` test class in `ca.uhn.fhir.jpa.dao.r4` includes several examples of tests that use custom SearchParameter definitions that you might be able to adapt (although I admit that I am having troubles finding examples that include `uri` parameters other than maybe cases that involve resources with extensions).

**Review by IanMMarshall — COMMENTED** [review]
Looks good. At minimum, I would ask that you rename the `theCompositList` parameter as per my comments. 

I also strongly recommend that we include tests for the `reference`, `number` and `uri` composite parameters if we are indeed going to support these as part of the sort criteria.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 3]
```diff
@@ -528,9 +528,25 @@ private void createSort(QueryStack theQueryStack, SortSpec theSort) {
 					break;
 				case QUANTITY:
 					theQueryStack.addSortOnQuantity(myResourceName, theSort.getParamName(), ascending);
+					break;	
+				case COMPOSITE:
+					List<RuntimeSearchParam> theCompositList = param.getCompositeOf();
```
[the remark this answers is not in the document]
Comment (frankjtao) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:533]: will rename to compositeList

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java (hunk 1 above)
Comment (frankjtao) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:571]: Will remove the other reference, number and uri option of the part of the composite, same as 
https://github.com/hapifhir/hapi-fhir/blob/d99b67820a402b438c3105214664bbd5fe6558d5/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java#L299

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java (hunk 2 above)
Comment (frankjtao) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:573]: Same as above

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java (hunk 1 above)
Comment (IanMMarshall) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java]: Looks good.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java (hunk 2 above)
Comment (IanMMarshall) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java]: Fair enough.

**Review by IanMMarshall — APPROVED** [review]
Looks good. Approved pending resolution of any merge conflicts.
